trl/trainer/utils.py

```diff
@@ -50,6 +50,59 @@ def update(self, current, n_steps):
         pass
 
 
+class DataCollatorForChatCompletionOnlyLM(DataCollatorForLanguageModeling):
+    def __init__(self, human_template, assistant_template, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.human_template = human_template
+        self.assistant_template = assistant_template
+
+    def torch_call(self, examples: List[Union[List[int], Any, Dict[str, Any]]]) -> Dict[str, Any]:
+        batch = super().torch_call(examples)
+
+        human_token_ids = self.tokenizer.encode(self.human_template, add_special_tokens=False)
+        assistant_token_ids = self.tokenizer.encode(self.assistant_template, add_special_tokens=False)
+        
+        labels = batch["labels"].clone()
+
+        for i in range(len(examples)):
+            assistant_token_ids_idxs = []
+            human_token_ids_idxs = []
+
+            for assistant_idx in np.where(batch["labels"][i] == assistant_token_ids[0])[0]:
+                # find the indexes of the start of an assistant answer.
+                if assistant_token_ids == examples[i]["input_ids"][assistant_idx : assistant_idx + len(assistant_token_ids)]:
+                    assistant_token_ids_idxs.append(assistant_idx + len(assistant_token_ids))
```

Comment:
Shouldn't this just be `append(assistant_idx)` (so that the `<|assistant|>` part is also ignored)?